parser/prism/Translator.cc

```diff
@@ -5079,7 +5097,32 @@ NodeVec Translator::translateEnsure(pm_begin_node *beginNode) {
 
             auto loc = translateLoc(start, end);
 
-            translatedEnsure = make_unique<parser::Ensure>(loc, move(bodyNode), move(ensureBody));
+            if (!directlyDesugar || !hasExpr(bodyNode, ensureBody)) {
+                translatedEnsure = make_unique<parser::Ensure>(loc, move(bodyNode), move(ensureBody));
+            } else {
+                // Build ast::Rescue expression with ensure field set
+                // When there's no rescue clause, create a new Rescue with empty rescue cases
+                ast::ExpressionPtr bodyExpr;
+                if (bodyNode != nullptr) {
+                    bodyExpr = bodyNode->takeDesugaredExpr();
+                } else {
+                    bodyExpr = ast::MK::EmptyTree();
+                }
+
+                ast::ExpressionPtr ensureExpr;
+                if (ensureBody != nullptr) {
+                    ensureExpr = ensureBody->takeDesugaredExpr();
+                } else {
+                    ensureExpr = ast::MK::EmptyTree();
+                }
+
+                // Create ast::Rescue with empty rescue cases
+                ast::Rescue::RESCUE_CASE_store cases;
+                auto rescueExpr = ast::make_expression<ast::Rescue>(loc, move(bodyExpr), move(cases),
+                                                                    ast::MK::EmptyTree(), move(ensureExpr));
```

Comment:
Perhaps it's just a matter of personal preference, but I like using a var+name over a comment when possible

```suggestion
                auto emptyCases = ast::Rescue::RESCUE_CASE_store{};
                auto emptyElseClause = ast::MK::EmptyTree();
                auto rescueExpr = ast::make_expression<ast::Rescue>(loc, move(bodyExpr), move(emptyCases),
                                                                    move(emptyElseClause), move(ensureExpr));
```